### PR TITLE
Cranelift: x64, aarch64, s390x, riscv64: ensure addresses are I64s.

### PR DESCRIPTION
**Pull request by cfallin**
@avanhatt has been looking at our address-mode lowering and found an example where when feeding an `I32`-typed address into a load or store, we can violate assumptions and get incorrect codegen.

This should never be reachable in practice, because all producers on 64-bit architectures use 64-bit types for addresses. However, our IR is insufficiently constrained, and allows loads/stores to `I32` addresses as well. This is nonsensical on a 64-bit architecture.

Initially I had thought we should tighten either the instruction definition's accepted types, or the CLIF verifier, to reject this. However both are target-independent, and we don't want to bake an assumption of 64-bit-ness into the compiler core. Instead this PR tightens specific backends' lowerings to rejecct loads/stores of `I32`-typed addresses.

aarch64 still has `lower_address` in Rust and no clear single place to put an LHS constraint, so I opted for an assert in the external ctor. riscv64 has no single point where we have a `Value` and produce an address (`gen_load` takes a `Reg`) so all load/store lowerings get the constraint.

tl;dr: no security implications as all producers use I64-typed addresses (and must, for correct operation); but we currently accept I32-typed addresses too, and this breaks other assumptions.

### CODE REVIEW

**Review by fitzgen — APPROVED** [review]

**Comment by cfallin** [comment]
(force-push to update commit description to remove redundant paragraph; no other changes)

**Comment by cfallin** [comment]
Updated to allow `R64` types as well (it seems that `table_*` ops rely on this). Also added the same check to `PairAMode` generation on aarch64.

File: cranelift/codegen/src/isa/riscv64/lower.isle [hunk 1]
```diff
@@ -697,63 +697,63 @@
 
 ;;;;;  Rules for `uload8`;;;;;;;;;
 (rule
-  (lower (uload8 flags p offset))
+  (lower (uload8 flags p @ (value_type (ty_64 _)) offset))
```
Comment (fitzgen) [inline cranelift/codegen/src/isa/riscv64/lower.isle:700]: Does `ty_64` succeed for `f64`? I guess the CLIF verifier would presumably catch that one, but it may make sense to use a purpose built extractor that is called `is_address_type` or something like that.

**Review by fitzgen — APPROVED** [review]
r=me with small suggestion if you want it

File: cranelift/codegen/src/isa/riscv64/lower.isle (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/riscv64/lower.isle:700]: It looks like the instruction definition (the `iAddr` type set) includes only ints and reference types, so this case can't happen. I do like `is_address_type` though; I'll do that.

**Review by avanhatt — APPROVED** [review]
